src/osd/ReplicatedPG.cc

```diff
@@ -1869,19 +1871,40 @@ bool ReplicatedPG::maybe_handle_cache(OpRequestRef op,
     // Promote too?
     switch (pool.info.min_read_recency_for_promote) {
     case 0:
-      promote_object(obc, missing_oid, oloc, promote_op);
+      if (promote_op) {
+	// sync promote
+        promote_object(obc, missing_oid, oloc, promote_op);
+      } else {
+	// async promote
+	queue_async_promote(poid, oloc);
+	start_async_promote();
+      }
       break;
     case 1:
       // Check if in the current hit set
       if (in_hit_set) {
-	promote_object(obc, missing_oid, oloc, promote_op);
+        if (promote_op) {
+	  // sync promote
+	  promote_object(obc, missing_oid, oloc, promote_op);
+	} else {
+	  // async promote
+          queue_async_promote(poid, oloc);
+          start_async_promote();
+        }
```

Comment:
this block could go into a maybe_promote_object(...) so that it isn't duplicated 4x.  and the hobject_t poid above would go away (since it's only used by the else branch here)